### ISSUE TITLE
Update state from the rootReducer level

### ISSUE THREAD

**Issue by qn895** [ioof-holdings/redux-dynostore#28]
### Is it a bug, feature request or question?

Question: I have a globalAction that's similar to `{type: "MERGE_PAYLOAD", payload: {blah: "blah"}`. Its implementation is something like this

```javascript
//subspaced-component-reducer.js
  ["MERGE_PAYLOAD"]: (state, action) => {
    return {
      ...state,
      ...action.payload,
    };
  }
```

An example use case is to load the configuration/settings of the specific component the user has previously saved. While putting the above piece of code in every component works, we are trying to generalize the dynamic reducer code as much as possible to encourage reusability. One simple way to do this is to update the reducer on the root level. I can do this to the static reducers like below.

```javascript
const rootReducer = (state, action) => {
  if (action.type === "MERGE_PAYLOAD") {
    return { ...state, [action.key]: {...state[action.key], ...action.payload }};
  }
  return appReducer(state, action);
};

const store = createStore(rootReducer, enhancers);

```

As you can see, within the rootReducer, the `state` will not have access to the dynamic reducers. My question is, what would be the best way to implement this for the dynamic reducers? I have looked into middlewares (dispatch and getState) and dynamic enhancers but still haven't found a good way to implement something like this. The key is intercepts what the nextState should be if the `action.type` explicitly equals `MERGE_PAYLOAD`.

### Which package(s) does this involve?

Redux-subspace or redux dynostore core

### Your Setup

| package         | version(s)
| --------------- | ----------
| redux           | 4.0
| redux-dynostore | 2.0

**Comment by mpeyper** [comment]
Hi @qn895,

I don't think there is currently anything that provides this sort of functionality.

If reusability is the goal, you could just create a shared library that exports a higher-order reducer that handles the merge functionality for you, e.g.

```js
export default reducer => (state, action) => {
  const newState = reducer(state, action)
  return action.type === 'MERGE_PAYLOAD' ? { ...newState, ...action.payload } : newState
}
```

and then wrap your reducers in that before attaching them.

I could also see an option added (or an extension to v2's `stateHandler` functionality?) to give some kind of `rootReducer` customisatio, but I'm not entirely sure what the API would look like.  Any idea on how you might want to configure dynostore to do this?  I'm thinking something like:

```js
const mergePayloadEnhancer = reducer => (state, action) => {
  if (action.type === "MERGE_PAYLOAD") {
    return { ...state, [action.key]: {...state[action.key], ...action.payload }};
  }
  return reducer(state, action);
};

const store = createStore(appReducer, dynostore(dynamicReducers({ rootEnhancers: [mergePayloadEnhancer] })));
```

Essentially these are just higher-order reducers, but we will compose them together an wrap them around the root reducer after we build the dynamic reducers and merge them with the static reducers (e.g. somewhere [here before returning](https://github.com/ioof-holdings/redux-dynostore/blob/master/packages/redux-dynostore-core/src/reducers/createDynamicReducer.js#L66)).  This is just off the top of my head, so there is likely better ways to do this?

**Comment by qn895** [comment]
I think you're right, I'll try out the first solution with the higher order reducer as quick solution. I really like the idea of the rootEnhancers option, though I wonder if we should have something akin to what we have for middleware (applyToRoot, applyToNamespacedRoots, applyToChildren) as well? 

But then again, I guess it might not matter since it already intercepts the root reducer after the dynamic reducers are merged.

Thanks for your help @mpeyper

**Comment by mpeyper** [comment]
> I wonder if we should have something akin to what we have for middleware (applyToRoot, applyToNamespacedRoots, applyToChildren) as well?

Interesting... I'm not sure how it would know what "level" the reducer is at (they are fairly agnostic of where in the tree they live), but conceptually like the idea.

Anyway, let me know how you go with option 1 and if you want to explore this enhancement further.

**Comment by qn895** [comment]
Yeah I got the option 1 working pretty well for the all the dynamic components so no problem so far. It works something like this:

```javascript
// template-enhancers.js
const withTemplateManager = reducer => (state, action) => {
  if (action.type === "LOAD_VIEW_SETTINGS") {
    return {
      ...state,
      ...action.payload
    };
  }
  return reducer(state, action);
};

export default withTemplateManager;
```

```javascript
//dynamic-component1.js
export default dynamic(
  "dynamicComponent1",
  attachReducer(withTemplateManager(reducer)),
  subspaced()
)(DynamicComponent1);
```

Works well for my case, but I think being able to compose multiple reducer enhancers at the root level before returning is much better in the long run. I'll be happy to explore this option and help out if needed.

**Comment by qn895** [comment]
I'm gonna close this thread since we found several options that work. If anyone need to compose multiple reducer enhancers at the root level, please feel free to look at how I implemented it here #29
